test/libsolidity/SolidityNameAndTypeResolution.cpp

```diff
@@ -7409,6 +7409,31 @@ BOOST_AUTO_TEST_CASE(builtin_reject_gas)
 	CHECK_ERROR(text, TypeError, "Member \"gas\" not found or not visible after argument-dependent lookup");
 }
 
+BOOST_AUTO_TEST_CASE(gas_left)
+{
+	char const* text = R"(
+		contract C {
+			function f() public returns (uint256 val) { return msg.gas; }
+		}
+	)";
+	CHECK_SUCCESS(text);
+
+	text = R"(
+		contract C {
+			function f() public returns (uint256 val) { return gasleft(); }
+		}
+	)";
+	CHECK_SUCCESS(text);
+
+	text = R"(
+		pragma experimental "v0.5.0";
+		contract C {
+			function f() public returns (uint256 val) { return msg.gas; }
+		}
+	)";
+	CHECK_ERROR(text, TypeError, "Member \"gas\" not found or not visible after argument-dependent lookup in msg");
+}
+
```

Comment:
Please add a test where `gasleft` is a local variable or function and check that resolution works correctly.

It could also test what happens if the signature of the function is different and whether overload resolution is performed.

Reply:
Regarding overload resolution: what's the expected behaviour here?

I just checked the behaviour of the global function "now" and noticed, that if "now" is overwritten/shadowed by a local function, then the global "now" will be inaccessible, even if the local function has a different signature... so at the moment overload resolution does not seem to work in such cases.

As of now "gasleft" has the same behaviour. I'm about to add tests for shadowing gasleft with local functions/variables, but I'm not sure about overload resolution - should overload resolution find the global function, if a local function has a different signature? If so, this should probably also be adjusted for the case of "now"...

Reply:
Yes, it should be consistent with the other cases: Shadowing between scopes, overloading in the same scope.